File: ColorTV Demo Objective-C/ColorTV Demo Objective-C/VideoViewController.m
//
//  VideoViewController.m
//  ColorTV Demo Objective-C
//
//  Created by Jordan Jasinski on 15/03/2017.
//  Copyright © 2017 Jordan Jasinski. All rights reserved.
//

#import "VideoViewController.h"

#import "DemoPlayerVideoView.h"

#if TARGET_OS_TV
@import COLORAdFramework;
#else
@import COLORMobileAdFramework;
#endif

static NSString *const kPlaybackLikelyToKeepUpKeyPath = @"currentItem.playbackLikelyToKeepUp";
static NSString *const kRateKeyPath = @"rate";

@interface VideoViewController ()

@property (nonatomic, assign, readwrite) BOOL isVideoPlaying;
@property (nonatomic, strong) UIActivityIndicatorView *loadingIndicator;
@property (nonatomic, strong) UITapGestureRecognizer *playPauseButtonRecognizer;
@property (nonatomic, strong) COLORRecommendationViewController *recommendationVC;
@property (nonatomic, strong) COLORNextItemRecommendationViewController *nextItemVC;

@end

@implementation VideoViewController

- (void)loadView {
    self.view = [[DemoPlayerVideoView alloc] init];
    [self setupLoadingIndicator];
    [self setupPlayPauseButtonIfNeeded];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    [self addObservers];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    if(self.isVideoPlaying) {
        [self pause];
    }
    
    [self removeObservers];
}

#pragma mark - Observers

- (void)addObservers {
    [self.view addGestureRecognizer:self.playPauseButtonRecognizer];
    
    [self.videoPlayer addObserver:self
                       forKeyPath:kRateKeyPath
                          options:NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld
                          context:nil];
    [self.videoPlayer addObserver:self
                       forKeyPath:kPlaybackLikelyToKeepUpKeyPath
                          options:NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld
                          context:nil];
}

- (void)removeObservers {
    [self.view removeGestureRecognizer:self.playPauseButtonRecognizer];
    
    [self.videoPlayer removeObserver:self forKeyPath:kRateKeyPath];
    [self.videoPlayer removeObserver:self forKeyPath:kPlaybackLikelyToKeepUpKeyPath];
}

- (UITapGestureRecognizer *)playPauseButtonRecognizer {
    if (!_playPauseButtonRecognizer) {
        _playPauseButtonRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self
                                                                             action:@selector(playPauseButtonClicked)];
        _playPauseButtonRecognizer.allowedPressTypes = @[@(UIPressTypePlayPause)];
    }
    return _playPauseButtonRecognizer;
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    
    if([keyPath isEqualToString:kRateKeyPath]) {
        float rate = [change[@"new"] floatValue];
        BOOL recommendationsShouldBePresented = self.recommendationVC && (self.presentedViewController != self.recommendationVC);
        if(rate == 0 && recommendationsShouldBePresented) {
            [self presentViewController:self.recommendationVC animated:YES completion:nil];
        }
        if(self.videoPlayer.currentItem.isPlaybackLikelyToKeepUp) {
            [self.loadingIndicator stopAnimating];
        } else {
            [self.loadingIndicator startAnimating];
        }
    } else if([keyPath isEqualToString:kPlaybackLikelyToKeepUpKeyPath]) {
        BOOL likelyToKeepUp = [change[@"new"] boolValue];
        if(likelyToKeepUp) {
            [self.loadingIndicator stopAnimating];
        } else {
            [self.loadingIndicator startAnimating];
        }
    }
}

#pragma mark - View

- (AVPlayer *)videoPlayer {
    return ((AVPlayerLayer *)self.view.layer).player;
}

- (void)setupLoadingIndicator {
    self.loadingIndicator = [[UIActivityIndicatorView alloc] initWithFrame:CGRectZero];
    self.loadingIndicator.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.loadingIndicator];
    [self.loadingIndicator.widthAnchor constraintEqualToConstant:100.0f].active = YES;
    [self.loadingIndicator.heightAnchor constraintEqualToConstant:100.0f].active = YES;
    [self.loadingIndicator.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor].active = YES;
    [self.loadingIndicator.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor].active = YES;
    [self.loadingIndicator startAnimating];
}

- (void)setupPlayPauseButtonIfNeeded {
#if TARGET_OS_IOS
    UIButton *playPauseButton = [UIButton buttonWithType:UIButtonTypeCustom];
    UIImage *image = [UIImage imageNamed:@"pause"];
    [playPauseButton setImage:image forState:UIControlStateNormal];
    playPauseButton.backgroundColor = [UIColor clearColor];
    [playPauseButton addTarget:self action:@selector(playPauseButtonClicked) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:playPauseButton];
    playPauseButton.layer.cornerRadius = 35.0f;
    playPauseButton.translatesAutoresizingMaskIntoConstraints = NO;
    [playPauseButton.widthAnchor constraintEqualToConstant:70.0f].active = YES;
    [playPauseButton.heightAnchor constraintEqualToConstant:70.0f].active = YES;
    [playPauseButton.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:40.0f].active = YES;
    [playPauseButton.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor constant:-40.0f].active = YES;
#endif
}

#pragma mark - Video

- (void)loadVideoWithURL:(NSURL *)url andId:(NSString *)identifier {
    AVPlayerItem *videoItem = [AVPlayerItem playerItemWithURL:url];
    
    if(self.videoPlayer) {
        [self.videoPlayer replaceCurrentItemWithPlayerItem:videoItem];
    } else {
        ((DemoPlayerVideoView *)self.view).player = [AVPlayer playerWithPlayerItem:videoItem];
    }
    
    [[COLORAdController sharedAdController] contentRecommendationControllerForPlacement:COLORAdFrameworkPlacementVideoStart
                                                                             andVideoId:identifier
                                                                         withCompletion:^(COLORRecommendationViewController * _Nullable viewController, NSError * _Nullable error) {
         if (error) {
             NSLog(@"::>> ConRec ERROR: %@", error);
             return;
         }
         if (!viewController) {
             NSLog(@"::>> ConRec No RecommendationViewController");
             return;
         }
         
         viewController.itemSelected = ^(NSString *videoId, NSURL *videoURL, NSDictionary *clickParams) {
             dispatch_async(dispatch_get_main_queue(), ^{
                 [self dismissViewControllerAnimated:YES completion:^{
                     [self loadVideoWithURL:videoURL andId:videoId];
                     [self play];
                 }];
             });
         };
         
         viewController.contentRecommendationClosed = ^() {
             dispatch_async(dispatch_get_main_queue(), ^{
                 [self dismissViewControllerAnimated:YES completion:^{
                     [self play];
                 }];
             });
         };
         
         self.recommendationVC = viewController;
     }];
    
}

- (void)play {
    [[self videoPlayer] play];
    self.isVideoPlaying = YES;
}

- (void)pause {
    [[self videoPlayer] pause];
    self.isVideoPlaying = NO;
}

- (void)playPauseButtonClicked {
    if(self.isVideoPlaying) {
        [self pause];
    } else {
        [self play];
    }
}

@end

File: ColorTV Demo Swift/ColorTV Test App Swift/VideoViewController.swift
//
//  VideoViewController.swift
//  ColorTV Demo Swift
//
//  Created by Szymon Litak on 04/05/2017.
//  Copyright © 2017 Jordan Jasinski. All rights reserved.
//

import UIKit
import Foundation
import AVFoundation

import COLORAdFramework

class VideoViewController: UIViewController {
    
    let kPlaybackLikelyToKeepUpKeyPath = "currentItem.playbackLikelyToKeepUp"
    var isVideoPlaying: Bool = false
    
    var videoView: DemoPlayerVideoView {
        return view as! DemoPlayerVideoView
    }
    
    @IBOutlet weak var loadingIndicator: UIActivityIndicatorView!
    private var recommendationVC: COLORRecommendationViewController!
    private var nextItemVC: COLORNextItemRecommendationViewController!
    
    var videoPlayer: AVPlayer? {
        return videoView.playerLayer.player;
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        let playPauseBtnRecognizer = UITapGestureRecognizer(target: self, action: #selector(playPauseButtonClicked))
        playPauseBtnRecognizer.allowedPressTypes = [NSNumber(value: UIPressType.playPause.rawValue)]
        view.addGestureRecognizer(playPauseBtnRecognizer)
        view.backgroundColor = UIColor.black
        loadingIndicator.startAnimating()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        videoPlayer?.addObserver(self, forKeyPath: "rate", options: [.new, .old], context: nil)
        videoPlayer?.addObserver(self, forKeyPath: kPlaybackLikelyToKeepUpKeyPath, options: [.new, .old], context: nil)
    }
    
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        if (keyPath == "rate") {
            if let rate = change?[.newKey] as? Float, rate == 0 {
                if recommendationVC != nil, presentedViewController != recommendationVC {
                    present(recommendationVC, animated: true, completion: nil)
                }
                
                if let isLikely = videoPlayer?.currentItem?.isPlaybackLikelyToKeepUp, isLikely {
                    loadingIndicator.stopAnimating()
                } else {
                    loadingIndicator.startAnimating()
                }
            }
        } else if keyPath == kPlaybackLikelyToKeepUpKeyPath {
            if let likelyToKeepUp = change?[.newKey] as? Bool {
                likelyToKeepUp ? loadingIndicator.stopAnimating() : loadingIndicator.startAnimating()
            }
        }
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        self.videoPlayer?.removeObserver(self, forKeyPath: "rate")
        self.videoPlayer?.removeObserver(self, forKeyPath: kPlaybackLikelyToKeepUpKeyPath)
    }
    
    func loadVideo(with url: URL, identifier: String) {
        var videoItem = AVPlayerItem(url: url)
        
        if let videoPlayer = videoPlayer {
            videoPlayer.replaceCurrentItem(with: videoItem)
        } else {
            videoView.player = AVPlayer(playerItem: videoItem)
        }
        
        COLORAdController.sharedAdController().contentRecommendationController(forPlacement: COLORAdFrameworkPlacementVideoStart, andVideoId: identifier) {
            (vc: COLORRecommendationViewController?, error: Error?) in
            guard let vc = vc, error == nil else {
                NSLog("::>> ConRec ERROR: \(error?.localizedDescription ?? "Unrecognized error")")
                return
            }
            vc.itemSelected = {(videoId: String, videoURL: URL, clickParams: [AnyHashable : Any]?) in
                DispatchQueue.main.async {
                    self.dismiss(animated: true) {
                        self.loadVideo(with: videoURL, identifier: videoId)
                        self.play()
                    }
                }
            }
            vc.contentRecommendationClosed = {
                DispatchQueue.main.async {
                    self.dismiss(animated: true) {
                        self.play()
                    }
                }
            }
            self.recommendationVC = vc
        }
        play()
    }
    
    func play() {
        videoPlayer?.play()
        isVideoPlaying = true
    }
    
    func pause() {
        videoPlayer?.pause()
        isVideoPlaying = false
    }
    
    func playPauseButtonClicked(sender: Any) {
        isVideoPlaying ? pause() : play()
    }
}
